library/extractor/src/main/java/com/google/android/exoplayer2/extractor/mp4/FragmentedMp4Extractor.java

```diff
@@ -1407,23 +1408,45 @@ private boolean readSample(ExtractorInput input) throws IOException {
           processSeiNalUnitPayload =
               ceaTrackOutputs.length > 0
                   && NalUnitUtil.isNalUnitSei(track.format.sampleMimeType, nalPrefixData[4]);
+          processSpsNalUnitPayload =
+              NalUnitUtil.isNalUnitSps(track.format.sampleMimeType, nalPrefixData[4]);
           sampleBytesWritten += 5;
           sampleSize += nalUnitLengthFieldLengthDiff;
         } else {
           int writtenBytes;
-          if (processSeiNalUnitPayload) {
-            // Read and write the payload of the SEI NAL unit.
+          if (processSeiNalUnitPayload || processSpsNalUnitPayload) {
+            // We need to process the payload further, so read it into the temporary buffer
+            // before writing.
             nalBuffer.reset(sampleCurrentNalBytesRemaining);
             input.readFully(nalBuffer.getData(), 0, sampleCurrentNalBytesRemaining);
             output.sampleData(nalBuffer, sampleCurrentNalBytesRemaining);
             writtenBytes = sampleCurrentNalBytesRemaining;
-            // Unescape and process the SEI NAL unit.
-            int unescapedLength =
-                NalUnitUtil.unescapeStream(nalBuffer.getData(), nalBuffer.limit());
-            // If the format is H.265/HEVC the NAL unit header has two bytes so skip one more byte.
-            nalBuffer.setPosition(MimeTypes.VIDEO_H265.equals(track.format.sampleMimeType) ? 1 : 0);
-            nalBuffer.setLimit(unescapedLength);
-            CeaUtil.consume(sampleTimeUs, nalBuffer, ceaTrackOutputs);
+            if (processSeiNalUnitPayload) {
+              // Unescape and process the SEI NAL unit.
+              int unescapedLength =
+                  NalUnitUtil.unescapeStream(nalBuffer.getData(), nalBuffer.limit());
+              // If the format is H.265/HEVC the NAL unit header has two bytes so skip one more byte.
+              nalBuffer.setPosition(
+                  MimeTypes.VIDEO_H265.equals(track.format.sampleMimeType) ? 1 : 0);
+              nalBuffer.setLimit(unescapedLength);
+              CeaUtil.consume(sampleTimeUs, nalBuffer, ceaTrackOutputs);
+            } else if (processSpsNalUnitPayload) {
+              // Parse SPS and update the aspect ratio
+              float pixelWidthHeightRatio = track.format.pixelWidthHeightRatio;
+              if (MimeTypes.VIDEO_H264.equals(track.format.sampleMimeType)) {
+                NalUnitUtil.SpsData spsData = NalUnitUtil.parseSpsNalUnitPayload(
+                    nalBuffer.getData(), 0, nalBuffer.limit());
+                pixelWidthHeightRatio = spsData.pixelWidthHeightRatio;
+              } else if (MimeTypes.VIDEO_H265.equals(track.format.sampleMimeType)) {
+                // TODO: Why is offset 1 necessary?
```

Comment:
This is necessary because the payload has been read assuming a 1 byte NAL unit header, but in H265 the header is actually 1 bytes, so the first byte in nalBuffer is actually the second byte of the header.

What you're doing here is identical to what's already happening on L1430 above, where `nalBuffer.setPosition` is passed `1` for HEVC and `0` otherwise. There is probably a clearer way that this could be written in code, but you can gain confidence from the pre-existing case that this is the right thing to do!

Reply:
Ah, thank you. I didn't notice that the code in line 1430 is already inside the if for `processSeiNalUnitPayload` and therefore doesn't get executed when processing the SPS. Now it totally makes sense that the offset is needed.

I'm wondering if the whole
```java
// Unescape and process the SEI NAL unit.
int unescapedLength =
    NalUnitUtil.unescapeStream(nalBuffer.getData(), nalBuffer.limit());
// If the format is H.265/HEVC the NAL unit header has two bytes so skip one more byte.
nalBuffer.setPosition(
    MimeTypes.VIDEO_H265.equals(track.format.sampleMimeType) ? 1 : 0);
nalBuffer.setLimit(unescapedLength);
```
could be moved before line 1424 to deduplicate things. Assuming that the unescaping is needed/acceptable for SPS NAL units.